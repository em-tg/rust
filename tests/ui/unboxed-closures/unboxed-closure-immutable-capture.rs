//@ check-pass
// Test that even unboxed closures that are capable of mutating their
// environment cannot mutate captured variables that have not been
// declared mutable (#18335)

fn set(x: &mut usize) { *x = 0; }

fn main() {
    let x = 0;
    move || x = 1; //~ WARNING cannot assign
    move || set(&mut x); //~ WARNING cannot borrow
    move || x = 1; //~ WARNING cannot assign
    move || set(&mut x); //~ WARNING cannot borrow
    || x = 1; //~ WARNING cannot assign
    || set(&mut x); //~ WARNING cannot borrow
    || x = 1; //~ WARNING cannot assign
    || set(&mut x); //~ WARNING cannot borrow
}
