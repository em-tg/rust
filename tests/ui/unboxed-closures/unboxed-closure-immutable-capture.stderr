warning[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:10:13
   |
LL |     move || x = 1;
   |             ^^^^^ cannot assign
   |
   = note: `#[warn(mut_non_mut)]` on by default
help: consider changing this to be mutable
   |
LL |     let mut x = 0;
   |         +++

warning[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:11:17
   |
LL |     move || set(&mut x);
   |                 ^^^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
LL |     let mut x = 0;
   |         +++

warning[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:12:13
   |
LL |     move || x = 1;
   |             ^^^^^ cannot assign
   |
help: consider changing this to be mutable
   |
LL |     let mut x = 0;
   |         +++

warning[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:13:17
   |
LL |     move || set(&mut x);
   |                 ^^^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
LL |     let mut x = 0;
   |         +++

warning[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:14:8
   |
LL |     || x = 1;
   |        ^^^^^ cannot assign
   |
help: consider changing this to be mutable
   |
LL |     let mut x = 0;
   |         +++

warning[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:15:12
   |
LL |     || set(&mut x);
   |            ^^^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
LL |     let mut x = 0;
   |         +++

warning[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:16:8
   |
LL |     || x = 1;
   |        ^^^^^ cannot assign
   |
help: consider changing this to be mutable
   |
LL |     let mut x = 0;
   |         +++

warning[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:17:12
   |
LL |     || set(&mut x);
   |            ^^^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
LL |     let mut x = 0;
   |         +++

warning: 8 warnings emitted

Some errors have detailed explanations: E0594, E0596.
For more information about an error, try `rustc --explain E0594`.
